core/tests/performance_framework/perf_services.py

```diff
@@ -0,0 +1,237 @@
+            error_msg = 'Unsupported browser specified: %s' % browser
+            raise ValueError(error_msg)
+
+        self.server = self.driver = self.proxy = None
+
+    def get_metrics_from_har_dict(self, page_url):
+        """Returns a PageSessionMetrics domain object initialized using
+        HTTP Archive (HAR) dict for a given page URL.
+
+        The HAR dict is retrieved using the proxy server by recording the
+        communication between the client and server.
+        """
+        with Xvfb() as _:
+            self._setup_proxy()
+            self._setup_driver(use_proxy=True)
+
+            self.proxy.new_har(page_url, options={'captureHeaders': True})
+
+            self.driver.get(page_url)
+
+            self._wait_until_page_load_is_finished()
+
+        har_dict = self.proxy.har
+
+        self._stop_proxy()
+        self._stop_driver()
+
+        page_session_metrics = perf_domain.PageSessionMetrics(
+            page_session_stats=har_dict, page_session_timings=None)
+
+        return page_session_metrics
+
+    def get_metrics_from_har_dict_cached_state(self, page_url):
+        """Returns a PageSessionMetrics domain object initialized using
+        HTTP Archive (HAR) dict for a given page URL while simulating a
+        cached state i.e, a return user.
+
+        To get it we require the use of a proxy server as we need to record the
+        communication between server and client.
+        """
+        with Xvfb() as _:
+            self._setup_proxy()
+            self._setup_driver(use_proxy=True)
+
+            # Fetch the page once. This leads to caching of various resources.
+            self.driver.get(page_url)
+            self._wait_until_page_load_is_finished()
+
+            # Start recording har data for the next page fetch.
+            self.proxy.new_har(page_url, options={'captureHeaders': True})
+            self.driver.get(page_url)
+
+            self._wait_until_page_load_is_finished()
+
+        har_dict = self.proxy.har
+
+        self._stop_proxy()
+        self._stop_driver()
+
+        page_session_metrics = perf_domain.PageSessionMetrics(
+            page_session_stats=har_dict, page_session_timings=None)
+
+        return page_session_metrics
+
+    def get_metrics_from_session_timings(self, page_url):
+        """Returns a PageSessionMetrics domain object initialized using
+        page load timings for a page URL.
+        """
+        with Xvfb() as _:
+            self._setup_driver(use_proxy=False)
+
+            self.driver.get(page_url)
+            self._wait_until_page_load_is_finished()
+
+            page_session_timings = (
+                self.driver.execute_script("return window.performance"))
+
+        self._stop_driver()
+
+        page_session_metrics = perf_domain.PageSessionMetrics(
+            page_session_stats=None, page_session_timings=page_session_timings)
+
+        return page_session_metrics
+
+    def get_metrics_from_cached_session_timings(self, page_url):
```

Comment:
I worry it may be difficult to distinguish the meaning of this from the other methods, but only slightly worried. Either way, to stay consistent with my other suggestions this one might be named:

``get_page_session_timings_for_url``